### DISCUSSION TITLE
How to Manipulate Variables at Each Time Step?

### DISCUSSION THREAD

**Discussion by mcrossover97** [idaholab/moose#18905] [Q&A Modules: General] [answered]
Hi everyone! 

I am trying to solve a reactive transport problem where equilibrium calculations are performed after solving momentum and continuity equations. However, I don't have a clue how I should implement this. I guess I have to hardcode some stuff in the main.c file (I have done this before using OpenFOAM).

For example, I want to write a code that multiplies the concentration of aqueous species by 2 at each time step. How is that possible and where should I write such code?

Thanks for your help and support!

Best,
Mohammad

**Comment by GiudGiud** [comment, marked as answer]
Hello

Are aqueous species a nonlinear variable (one you solve for in the numerical system) ?
If not, then use an auxkernel to do those operations. 
If so, then you can either:
- fit all the operations during the timestep (both equilibrium and momentum/continuity) and solve for them together. This is what really happens in the system anyway, reactions and advection are happening at the same time
- use a transient multiapp setup and split both solves on each app, then use transfers to re-write the ICs to each app using the final state of the previous solve

Guillaume

**Reply by mcrossover97** [reply]
Thanks for your answer!

Yes, they are nonlinear variables and I can't use an auxkernel. I probably have to go with your second suggestion as I'm treating transfer and chemistry separately (It's a sequential approach). This usually works if your timesteps are small enough.
